### ISSUE TITLE
kpatch-build fails when patched module calls new vmlinux function

### ISSUE THREAD

**Issue by joe-lawrence** [dynup/kpatch#783]
7770c18f93cc breaks patches which add calls from patched modules to new functions in a patched vmlinux.  That change enabled undefined symbol checking of the build log by kpatch-build, previously these checks were silently failing.  The `module-call-external.patch` integration test now fails to build as it adds a new symbol to the kernel and calls it from the nfsd module.

There's a little bit of git history behind (allowing) undefined symbols:

- 2014-08-29 - 4dee89269c83 ("add support for shadow variables")
-- Adds "undefined reference" kbuild.log check
-- Adds `LDFLAGS_vmlinux="--warn-unresolved-symbols"` and `LDFLAGS_MODULE="--warn-unresolved-symbols"`

- 2014-10-01 - ea445c03338a ("allow shadow functions to be called from modules")
-- Replaces `LDFLAGS_MODULE` with `KBUILD_MODPOST_WARN=1`
-- Adds "undefined!" kbuild.log check

- 2015-10-03 - f5de932b8dcf ("allow patched modules to call external functions")
-- Modifies the kmod core module to look for "external" symbols, which may be exported (as previously checked before this change), or provided by another .o within the patch module
-- Adds `module-call-external.patch` to the integration tests

- 2016-12-05 - 1330b4a3e5e9 ("kpatch-build: fix 'undefined reference to kpatch_shadow_*' errors")
-- Moves `--warn-unresolved-symbols` linker option to kpatch-gcc

- 2017-11-09 - 7770c18f93cc ("kpatch-build: Fix the logfile checks for 'undefined'")
-- Makes the undefined symbol checks actually work

So I think the intention was to tolerate undefined symbols to support the shadow variable API as provided by the kpatch core module, as well as calling new vmlinux functions from patched modules.

Shadow variable API references from patched modules and patched vmlinux tested as OK.  However, I think we need to be a little more clever about how we handle cases like `module-call-external.patch`.  Perhaps we should collect the list of reported undefined symbols and compare to symbols provided by the patch before considering it a build failure?

edit: added info about f5de932 ("allow patched modules to call external functions")

**Comment by jpoimboe** [comment]
I don't remember this patch causing undefined symbol issues before.  During what time frame was the 'undefined reference' check broken?  I can see it was fixed with 7770c18, how long was it broken before that?

**Comment by sm00th** [comment]
In case we will be going with @joe-lawrence's suggestion from the first comment: https://github.com/sm00th/kpatch/tree/extsym contains a sketch of how it may look like.

**Comment by joe-lawrence** [comment]
Hi @sm00th , thanks for checking this out.  I think this could work, though what happens if the new function/object is declared as static?  (ie, what happens if the kpatch introduces a local symbol named FOO to module A, is it reported as new?)

Also, a really small nit, but I think this can be consolidated:
```
cat cdo_output.tmp | grep -e "new function:" -e "new object:" | \
			awk '{print $NF}' >>${TEMPDIR}/new_symbols
```
into one awk line like this (untested):
```
awk '/new function:/ || /new object:/ {print $NF}' cdo_output.tmp >>${TEMPDIR}/new_symbols

**Comment by sm00th** [comment]
> I think this could work, though what happens if the new function/object is declared as static?

Yes, it did treat static functions as new and didn't fail. I've changed it so that only symbols with STB_GLOBAL bind will be considered when checking undefined symbols. Do you think this will be enough?

> one awk line like this (untested)

That works, thank you.


I was also wondering if reporting new objects should be an option instead of default behavior or limited in some other way because I am getting a bunch of new object reports normal users won't care about:

```
...
af_netlink.o: new object: __warned.19015
af_netlink.o: new object: __warned.19017
af_netlink.o: new object: __warned.19089
af_netlink.o: new object: __warned.19040
af_netlink.o: new object: __warned.19081
af_netlink.o: new object: __warned.57878
...
```

**Comment by sm00th** [comment]
> On the other hand, maybe the same goal can be accomplished entirely inside kpatch-build

Oh, I do like this solution a lot. It makes things easier and we won't have to worry about updating hardcoded symbol names in kpatch-build if new APIs like shadow variable will be introduced to kpatch.ko.

I am a bit concerned about finding kpatch.ko's location though. Will trying to determine it through modinfo and then falling back to assumption that we are running from a git tree be enough? Right now the function doing this looks [like this](https://github.com/sm00th/kpatch/commit/db1aee7ef6a85f7349de498318685c2fce12a185#diff-c4bc518fff9b8f8607f5f407c34c8809R190).

**Comment by euspectre** [comment]
Thanks for working on this!

In fact, the issue shows up not just with the tests. The fix for CVE-2016-7097, for example, does exactly the same as descibed above: `posix_acl_update_mode` is added in vmlinux and called from several file system modules.

The build of the binary patch with the newer kpatch-build failed silently, and it took a while till I understood what went wrong there.

**Comment by jpoimboe** [comment]
> On the other hand, maybe the same goal can be accomplished entirely inside kpatch-build
> 
> Oh, I do like this solution a lot. It makes things easier and we won't have to worry about updating hardcoded symbol names in kpatch-build if new APIs like shadow variable will be introduced to kpatch.ko.
> 
> I am a bit concerned about finding kpatch.ko's location though. Will trying to determine it through modinfo and then falling back to assumption that we are running from a git tree be enough? Right now the function doing this looks like this.

I also like this solution, though for the global kpatch.ko functions, it might be less fragile if we just hard code the function names.

**Comment by joe-lawrence** [comment]
> I am a bit concerned about finding kpatch.ko's location though. Will trying to determine it through modinfo and then falling back to assumption that we are running from a git tree be enough?

Since the typical use case is going to be loading kpatches via the kpatch utility, you could steal its `find_core_module()`.  Some of those search paths might be now deprecated, but at least that way kpatch-build would be looking in the same place as the kpatch utility.

From the latest patch:
> ```
> +for ur in $(cat ${TEMPDIR}/undefined_references); do
> +	grep -q "^${ur}$" ${TEMPDIR}/new_symbols || die "Undefined reference: ${ur}"
> +done
> ```

if the two files were sorted, you could use the `comm` utility to dump out the entire final list of undefined symbols:
```
UNDEFINED=$(comm -23 undefined_references new_symbols | tr '\n' ' ')
[[ ! -z "$UNDEFINED" ]] && die "Undefined symbols: $UNDEFINED"
```
though the `comm` invocation would probably need a comment to explain what '-23' means and you'd end up with just as many lines.  Keep the for...loop if you prefer.

**Comment by sm00th** [comment]
> I also like this solution, though for the global kpatch.ko functions, it might be less fragile if we just hard code the function names.

If there won't be any outburst of new exported functions in kpatch.ko it should be ok. But I think we'll have to maintain a proper function list, `grep -qv kpatch_shadow` that is there now can (unlikely, but still possible) mask real issues.

> Since the typical use case is going to be loading kpatches via the kpatch utility, you could steal its find_core_module().

Ah, I thought there might be something already implemented. Although just bluntly copying it over might not be such a good idea and introducing a utility library for just this is an overkill. Whats your stance on hardcoding the function list?

> you could use the comm utility to dump out the entire final list of undefined symbols

I think printing the full list of undefined symbols is an important enough benefit to switch to this.

**Comment by joe-lawrence** [comment]
> Whats your stance on hardcoding the function list?

Checking kpatch.ko dynamically saves us the future headache of new/renamed symbols... this is more complicated than a static list, but I think the accuracy outweighs the script-code costs.  /two cents  Maybe there are other complications that @jpoimboe sees here?

**Comment by jpoimboe** [comment]
The location of the core module could be dependent on the distro setup.  From past experience it seems like different people build/install the core module in different ways/places.  Also, if kpatch-build is in its own RPM, for example, kpatch.ko might not even exist on the system at all.

And also, it just adds to the complexity of kpatch-build by adding another dependency.  Currently it's completely ignorant of kpatch.ko, which is nice.  Maybe I'm just being pessimistic, but adding that dependency could bring in more bugs that we can't yet envision.

I think hard-coding the list of needed symbols is fine.  Livepatch is the future, and kpatch.ko is pretty much deprecated at this point.  Hopefully we're done adding new features to it.

**Comment by joe-lawrence** [comment]
Thanks for elaborating .. these are good points, especially the part about not adding many more kpatch features :)

**Comment by sm00th** [comment]
Updated the branch with 'comm' and hardcoded shadow_var symbol list.

**Comment by jpoimboe** [comment]
@sm00th It's looking good to me, can you open a pull request?  Add a "Fixes #783." line to the commit msg.
